File: packages/core/src/ratelimit/lua/fixedWindow.lua
-- KEYS[1] = key (hash: windowStart, count)
-- ARGV[1] = limit, ARGV[2] = windowMs, ARGV[3] = capacity (unused), ARGV[4] = now, ARGV[5] = cost
local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local now = tonumber(ARGV[4])
local cost = tonumber(ARGV[5])

local windowStart = math.floor(now / windowMs) * windowMs
local raw = redis.call('HMGET', KEYS[1], 'windowStart', 'count')
local storedWindowStart = tonumber(raw[1])
local count = tonumber(raw[2]) or 0

if storedWindowStart ~= windowStart then
  count = 0
end

local projected = count + cost
local allowed = projected <= limit
if allowed then
  count = projected
end

redis.call('HSET', KEYS[1], 'windowStart', windowStart, 'count', count)
redis.call('PEXPIRE', KEYS[1], windowMs)

local retryAfter = 0
if not allowed then
  retryAfter = windowStart + windowMs - now
end

return { allowed and 1 or 0, math.max(0, limit - count), retryAfter }

File: packages/core/src/ratelimit/algorithms/fixedWindow.ts
import type { Policy } from '../stores/Store.js';
import type { AlgorithmOutput } from './types.js';

export interface FixedWindowState {
  readonly windowStart: number;
  readonly count: number;
}

/**
 * Sabit aralıkta sayaç. Pencere sınırında burst'e izin verir (bkz. PLAN.md §4)
 * ama en basit ve en ucuz algoritma.
 */
export function fixedWindow(
  state: FixedWindowState | undefined,
  policy: Policy,
  now: number,
  cost: number,
): AlgorithmOutput<FixedWindowState> {
  const windowStart = Math.floor(now / policy.windowMs) * policy.windowMs;
  const currentCount = state && state.windowStart === windowStart ? state.count : 0;
  const projected = currentCount + cost;
  const allowed = projected <= policy.limit;
  const count = allowed ? projected : currentCount;

  return {
    state: { windowStart, count },
    result: {
      allowed,
      limit: policy.limit,
      remaining: Math.max(0, policy.limit - count),
      retryAfterMs: allowed ? 0 : windowStart + policy.windowMs - now,
    },
  };
}
